### PR TITLE
Redirector Terraform

### LINKED ISSUE #549: ENS protocol redirect server
### TL;DR
A prototype for this exists in PR #546 

The current [ens protocol spec](https://github.com/google/exposure-notifications-verification-server/blob/main/docs/ens-spec.md) is not clickable in all messaging apps that may be used to receive SMS deep links on mobile devices.

In order to ensure that all users have a clickable link, we will provide a redirect server that verification server operators can use to 

### Design

**Proposal**

Simple URL redirect service. This needs to support custom subdomain mapping to ENS link.

Having unique subdomains allows per-region applications to to register that intent and handle it in the mobile OS.

As an example.

https://us-wa.example.com/v?c=1234abcd5678efgh

will redirect to 

ens://v?c=1234abcd5678efgh&r=US-WA

Configuration will consist of 2 elements
URI prefix to simply rewrite to ens://
and the region to append
will add "&r=REGION"

redirect requests that reach the server and do not match will return a 404.

**Alternatives considered**

the existing ens:// schema has proven not to work. 

**Resources**

[ens protocol spec](https://github.com/google/exposure-notifications-verification-server/blob/main/docs/ens-spec.md)

**Additional information**

For production deployment, request logging should be disabled.

/kind enhancement
/assign

### PR DESCRIPTION
**Pull request by icco**
Continues #549, does not include DNS changes.

**Release Note**

```release-note
Adds redirector to terraform services
```

### CODE REVIEW

File: terraform/service_enx_redirect.tf [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  name     = "enx-redirect"
+  project  = var.project
+
+  backend {
+    group = google_compute_region_network_endpoint_group.enx-redirect.id
+  }
+}
+
+resource "google_cloud_run_domain_mapping" "enx-redirect" {
+  for_each = var.enx-redirect_custom_domains
```
Comment (sethvargo) [inline terraform/service_enx_redirect.tf:133]: I don't think this variable is defined?

File: builders/build.yaml [hunk 2]
```diff
@@ -166,6 +166,33 @@ steps:
...
+  args:
+  - 'build'
+  - '--file=builders/service.dockerfile'
+  - '--tag=gcr.io/${PROJECT_ID}/${_REPO}/enx-redirect:${_TAG}'
+  - '--build-arg=SERVICE=enx-redirect'
+  - '.'
+  waitFor:
+  - 'build-enx-redirect'
+
+
```
Comment (sethvargo) [inline builders/build.yaml:195]: Nit: too many NL

**Comment by icco** [comment]
Don't want to merge until @mikehelmick's PRs are in.

/hold
/assign sethvargo

File: terraform/variables.tf [hunk 3]
```diff
@@ -178,6 +178,12 @@ variable "adminapi-host" {
   description = "Domain adminapi is hosted on."
 }
 
+variable "redirect_domain_map" {
+  type        = map(string)
```
Comment (sethvargo) [inline terraform/variables.tf:182]: ```object(code=string, host=string)```

File: terraform/service_redirect.tf [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        resources {
+          limits = {
+            cpu    = "1"
+            memory = "512Mi"
+          }
+        }
+
+        dynamic "env" {
+          for_each = merge(
+            local.cache_config,
```
[the remark this answers is not in the document]
Comment (icco) [reply terraform/service_redirect.tf]: Nope, removed.

**Review by sethvargo — APPROVED** [review]
/lgtm

**Pull request merged**